Search coordinator uses event.ingested in cluster state to do rewrites

quux00 opened the pull request:
Min/max range for the event.ingested timestamp field (part of Elastic Common
Schema) was added to IndexMetadata in cluster state for searchable snapshots
in https://github.com/elastic/elasticsearch/pull/106252.

This commit modifies the search coordinator to rewrite searches to MatchNone
if the query searches a range of event.ingested that, from the min/max range
in cluster state, is known to not overlap. This is the same behavior we currently
have for the `@timestamp` field.

quux00 committed:
Search coordinator uses event.ingested in cluster state to do rewrites

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Update docs/changelog/110352.yaml

quux00 committed:
code and comments cleanup for review

javanna commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -1854,14 +1860,15 @@ public boolean allPendingDanglingIndicesWritten() {
     @Nullable
-    public DateFieldMapper.DateFieldType getTimestampFieldType(Index index) {
-        return timestampFieldMapperService.getTimestampFieldType(index);
+    public Map<String, DateFieldMapper.DateFieldType> getTimestampFieldTypeMap(Index index) {
```
Could we avoid having a map, especially as we know we support having only max two entries? I think that this signature is more generic than it needs to be, and it requires additional knowledge / context when looking at the code.

javanna commented on server/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java:
```diff
@@ -436,11 +436,11 @@ public String getWriteableName() {
         final MappedFieldType fieldType = coordinatorRewriteContext.getFieldType(fieldName);
         if (fieldType instanceof final DateFieldMapper.DateFieldType dateFieldType) {
-            if (coordinatorRewriteContext.hasTimestampData() == false) {
+            if (coordinatorRewriteContext.hasTimestampData(fieldName) == false) {
```
do range queries on date fields that are not `@timestamp` or `event.ingested` still work? I see that `hasTimestampData` throws exception? 

On a similar note, I wonder if those new methods exposed in the context class are too broad in that they take a generic field name string as an argument, but after all support only two possible values for it. Should we then have two explicit separate methods for these? That though would require moving the if to the caller code. I don't have a strong opinion on this, but it feels like at least the getMin and getMax should not throw IAE and return some placeholder like before for other date fields and the previous logic still works as before?

quux00 replied:
Once the refactoring for removing the `Map<String, DateFieldMapper.DateFieldType>` was done (in [this commit](https://github.com/elastic/elasticsearch/pull/110352/commits/005ed5035080791e91baaff2f0313123830b10f2)), I then started work to change this class, but I ended up not making any changes here as I think it is coded right. 

First, the `getRelation` call is only called in during a coordinator rewrite. 

Second `coordinatorRewriteContext.getFieldType` returns `null` if you pass in a fieldName that is not supported for coordinator rewrites. If you get back null, the `if` check on the next line will return false and none of the rest of the code executes - we just return `MappedFieldType.Relation.INTERSECTS`.

The other calls (which can throw IllegalArgumentException now) are all guarded by that if-check.

It also feels wrong to put separate `getMin` and `getMax` methods for `@timestamp` and `event.ingested` because that means the RangeQueryBuilder has to know (and hardcode) knowledge of what timestamps are in cluster state (law of Demeter violation). The `RangeQueryBuilder` just knows what `fieldName` it is querying against and that's all it needs to know. It asks the `coordinatorRewriteContext` whether it supports that field and has useful min/max ranges to return.

javanna replied:
I agree. I missed that getFieldType returns null . You could slightly adjust `CoordinatorRewriteContext#getFieldType` to return a `DateFieldType` and replace the instanceof with a null check, to make this all more explicit. it is though odd that we return the field type, but we never really do anything with it. Up to you how you want to clean this up a little.

javanna reviewed (commented):
I left a couple of comments on internal API design etc.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search-coordinator/uses-event.ingested-in-cluster-state

quux00 committed:
Modify CoordinatorRewriteContextProvider time series code hook

It now looks up time range from timeseries if the @timestamp range from cluster
state is not useful (UKNOWN) or not ready (not all shards accounted for).
It no longer also depends on eventIngested being in a non-ready/non-usable state.

quux00 commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java:
```diff
@@ -49,18 +52,30 @@ public CoordinatorRewriteContext getCoordinatorRewriteContext(Index index) {
+            // 3) there is no time series timestamp range present
             if (timestampRange == null) {
                 return null;
             }
```
Thanks for clarifying @martijnvg. I've made what I think are the appropriate changes in [this new commit](https://github.com/elastic/elasticsearch/pull/110352/commits/c15067243e783bd5629cf7ac3a739f5b5691ff4e). 

Let me know if that looks right to you now.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search-coordinator/uses-event.ingested-in-cluster-state

quux00 committed:
Created CachedTimestampFieldInfo and modified CoordinatorRewriteContext and replaced Map<String, DataFieldType> with it

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search-coordinator/uses-event.ingested-in-cluster-state

quux00 committed:
spotlessApply fix

javanna commented on server/src/main/java/org/elasticsearch/indices/CachedTimestampFieldInfo.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Data holder of timestamp fields held in cluster state IndexMetadata.
+ */
+public class CachedTimestampFieldInfo {
```
make it final? Perhaps I would call it `DateFieldRange`. Did you think it's important to include "Cached" in the name? Also Timestamp has a reference to `@timestamp` that maybe we should loosen up? (not a big deal, curious to know what you think)

quux00 replied:
Done. Changed class name to `DateFieldRangeInfo`. I didn't want to use just `DateFieldRange` since that implies a range only and this object is used for the field types as well than the ranges. But happy to change it if you don't like the new name.

javanna replied:
it's fine!

javanna commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java:
```diff
@@ -49,18 +48,25 @@ public CoordinatorRewriteContext getCoordinatorRewriteContext(Index index) {
+
+        // ensure the cached info has the latest ranges from cluster state
+        timestampsFieldInfo.setTimestampRange(indexMetadata.getTimestampRange());
+        timestampsFieldInfo.setEventIngestedRange(indexMetadata.getEventIngestedRange());
```
I left a question below as well on this new class. I wonder if these set calls are a bit of a smell. Maybe it feels odd that the mapping supplier returns a timestamp field info object, that needs updating. Is that a signal that maybe we should keep field types and ranges separate, perhaps have a compound field type abstraction and a compound field range abstraction, similar to what we had before but supporting more than a single field?

quux00 replied:
Changed. See above comment.

javanna reviewed (commented):
thanks, almost there, a couple more questions.

quux00 commented on server/src/main/java/org/elasticsearch/indices/CachedTimestampFieldInfo.java:
```diff
@@ -0,0 +1,67 @@
+
+    public void setEventIngestedRange(IndexLongFieldRange eventIngestedRange) {
+        this.eventIngestedRange = eventIngestedRange;
+    }
```
No. I added them because it felt weird for the `TimeFieldMapperService` to only fill in the field types and not the ranges on these objects, but then I didn't know if the ranges it filled in would potentially be stale vs. what is in cluster state.

In any case, I agree this is even more awkward so I've changed it to being immutable and `TimeFieldMapperService` to only fill in the field types and not the ranges.

quux00 committed:
PR feedback - changed CachedTimestampFieldInfo to DateFieldRangeInfo and made it immutable

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search-coordinator/uses-event.ingested-in-cluster-state

javanna commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContext.java:
```diff
@@ -53,29 +61,98 @@ public CoordinatorRewriteContext(
+     * @return min timestamp for the field from IndexMetadata in cluster state.
+     */
+    long getMinTimestamp(String fieldName) {
+        if (fieldName.equals(DataStream.TIMESTAMP_FIELD_NAME)) {
```
super minor nitpick: fieldName can never be null right? I see that getMaxTimestamp does the equal the other way around, that's not going to throw NPE, while this one is different. Perhaps it does not matter, but I wondered why.

quux00 replied:
Thanks, good catch. I'll make them consistent.

javanna reviewed (approved):
LGTM (left a small question, no need for another review round)

quux00 commented:
buildkite test this

quux00 committed:
Merge remote-tracking branch 'elastic/main' into search-coordinator/uses-event.ingested-in-cluster-state

quux00 committed:
minor change based on PR feedback

martijnvg reviewed (commented):
LGTM2 (I thought I clicked submit review)
